### PR TITLE
387/new unit testing

### PR DESCRIPTION
**Pull request by davbyr**
Brand new SHINY unit testing system utilising python's unittest library. Main changes:

1. unit_test.pt completely replaced with a file that controls access to all the different testing files.
2. Old unit test file copied to unit_test_OLD for nostalgic reasons
3. Each section of the original unit test has been moved into its own file, beginning with test_
4. New file for containing file paths (which have been tidied a bit). This is called unit_test_files.py. If you look inside the test files you will see that I have a line calling `import unit_test_files as files`. Then you can access whichever file you need with, e.g. `files.fn_nemo_dat`.
5. A new script for auto-generating a unit_test_contents file called generate_unit_test_contents.py. Will search though all test modules and print modules/methods to file.

I've included plenty of commenting in the files so hopefully that should help clear up how the new system works. There is also a file called test_TEMPLATE.py to show how to set up a new testing file.

Using this new system, unit_test.py simply imports the different test files, adds them to a thing called a testSuite and then runs the suite. I've hopefully made this file fairly flexible so you can pick and choose which tests you want to run (if you want to).

AC:
New unit_test.py runs successfully from the usual COAsT directory.
You are happy with the changes I have made.

As Jeff pointed out this morning, all of the old messages have been removed from the unit test now, but it should always be clear where the problem is if the units are designed correctly. Please take a look and double check I have transferred your units correctly. If you want to add any extra messages, go ahead and do it. A good place to add messages is in the self.assertTrue statements at the end of each unit.

This is a big change so feel free to take some time to mull it over.

### CODE REVIEW

**Review by anwiseNOCL — APPROVED** [review]
This is definitely a more scalable solution. Nice work!

I think this is good to merge as is, with the knowledge that we might need to tweak things in subsequent pull request. 

For example, one of the contour tests is failing now (see below, i think i must have broken it in my last pull request but some how it didn't show up). I will fix that in my next pull request rather than within this one. But there seem to be a couple others as well. Do you get there Fails?

```
======================================================================
ERROR: test_calculate_preddure_gradient_griven_flow (test_isobath_contour_methods.test_isobath_contour_methods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/users/anwise/COAsT_v1/unit_testing/test_isobath_contour_methods.py", line 83, in test_calculate_preddure_gradient_griven_flow
    cont_f.calc_geostrophic_flow(nemo_t, config_u=files.fn_config_u_grid,
  File "/home/users/anwise/COAsT_v1/coast/contour.py", line 726, in calc_geostrophic_flow
    self._update_cross_flow_vars(
  File "/home/users/anwise/COAsT_v1/coast/contour.py", line 442, in _update_cross_flow_vars
    self.data_cross_flow[var] = tmp_var[:-1]
  File "/home/users/anwise/miniconda3/envs/coast_v127/lib/python3.8/site-packages/xarray/core/dataset.py", line 1564, in __setitem__
    self.update({key: value})
  File "/home/users/anwise/miniconda3/envs/coast_v127/lib/python3.8/site-packages/xarray/core/dataset.py", line 4253, in update
    merge_result = dataset_update_method(self, other)
  File "/home/users/anwise/miniconda3/envs/coast_v127/lib/python3.8/site-packages/xarray/core/merge.py", line 988, in dataset_update_method
    return merge_core(
  File "/home/users/anwise/miniconda3/envs/coast_v127/lib/python3.8/site-packages/xarray/core/merge.py", line 651, in merge_core
    raise MergeError(
xarray.core.merge.MergeError: unable to determine if these variables should be coordinates or not in the merged result: {'longitude'}

======================================================================
FAIL: test_tidegauge_for_tabulated_data (test_tidegauge_methods.test_tidegauge_methods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/users/anwise/COAsT_v1/unit_testing/test_tidegauge_methods.py", line 160, in test_tidegauge_for_tabulated_data
    self.assertTrue(check1, 'check1')
AssertionError: False is not true : check1

======================================================================
FAIL: test_compare_processed_profile_with_model (test_profile_methods.test_profile_methods) [Gridded obs_operator]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/users/anwise/COAsT_v1/unit_testing/test_profile_methods.py", line 64, in test_compare_processed_profile_with_model
    self.assertTrue(check3, 'check3')
AssertionError: False is not true : check3

----------------------------------------------------------------------
Ran 60 tests in 339.379s

FAILED (failures=2, errors=1)
```

**Comment by davbyr** [comment]
@anwiseNOCL interesting. Just tried again and I do get one error now:
```
======================================================================
ERROR: test_calculate_preddure_gradient_griven_flow (test_isobath_contour_methods.test_isobath_contour_methods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dbyrne/code/COAsT/./unit_testing/test_isobath_contour_methods.py", line 83, in test_calculate_preddure_gradient_griven_flow
    cont_f.calc_geostrophic_flow(nemo_t, config_u=files.fn_config_u_grid,
  File "/Users/dbyrne/code/COAsT/coast/contour.py", line 726, in calc_geostrophic_flow
    self._update_cross_flow_vars(
  File "/Users/dbyrne/code/COAsT/coast/contour.py", line 442, in _update_cross_flow_vars
    self.data_cross_flow[var] = tmp_var[:-1]
  File "/Users/dbyrne/opt/anaconda3/envs/workshop040222/lib/python3.8/site-packages/xarray/core/dataset.py", line 1563, in __setitem__
    self.update({key: value})
  File "/Users/dbyrne/opt/anaconda3/envs/workshop040222/lib/python3.8/site-packages/xarray/core/dataset.py", line 4252, in update
    merge_result = dataset_update_method(self, other)
  File "/Users/dbyrne/opt/anaconda3/envs/workshop040222/lib/python3.8/site-packages/xarray/core/merge.py", line 988, in dataset_update_method
    return merge_core(
  File "/Users/dbyrne/opt/anaconda3/envs/workshop040222/lib/python3.8/site-packages/xarray/core/merge.py", line 651, in merge_core
    raise MergeError(
xarray.core.merge.MergeError: unable to determine if these variables should be coordinates or not in the merged result: {'longitude'}

----------------------------------------------------------------------
Ran 60 tests in 255.738s

FAILED (errors=1)
```

I don't get the profile or tidegauge one though, which is interesting. This DID work before I merged develop back into this branch, but I didnt test afterwards. So something must have changed in this merge. 

I think its probably a good idea to make sure the whole unit test is working before merging this branch into develop, so I'll take a look at these errors.

**Comment by anwiseNOCL** [comment]
@davbyr . Ok but don't worry about the contour one as I have sorted it in a branch that I will pull after (Even though it is weird that this error didn't show up before, the code does need to be made more robust so this error can't happen).

For the tide gauge Fail. It is the same issue I had with the previous unit test item 7m.

For the profile one, it is failing because probably need to use assertAlmostEqual(a, b) or np.isclose. i.e. that float evaluation can be slightly off on different machine given the precision of the input data compared to that float.

**Comment by davbyr** [comment]
@anwiseNOCL Good point yeah. I've changed the == to np.isclose in test_profile_methods(). So maybe these tests should be OK for you now.

**Comment by anwiseNOCL** [comment]
@davbyr Checked again and the Profiles error is now gone.

File: unit_testing/test_tidegauge_methods.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            check1 = len(tidegauge_list) == 2
+            # TEST: Check lowestoft matches
+            check2 = all(tidegauge_list[1].dataset == lowestoft.dataset)
+
+            self.assertTrue(check1, "check1")
+            self.assertTrue(check2, "check2")
+
+        with self.subTest("Plot multiple gauge"):
+            f, a = coast.Tidegauge.plot_on_map_multiple(tidegauge_list)
+            f.savefig(files.dn_fig + "tidegauge_multiple_map.png")
```
Comment (jpolton) [inline unit_testing/test_tidegauge_methods.py:129]: The figures has "grid on" or something equivalent, that looks a bit weird as it overlays grid line in the rectangular region spanning the two Lowestoft and Llandudno sites
tidegauge_multiple_map
.

File: unit_testing/test_tidegauge_methods.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+            # TEST: Check attribute dictionary and length of sea_level.
+            check1 = len(lowestoft.dataset.sea_level) == 193
+            check2 = lowestoft.dataset.attrs == test_attrs
+            self.assertTrue(check1, "check1")
+            self.assertTrue(check2, "check2")
+
+        with self.subTest("Plot Single GESLA"):
+            f, a = lowestoft.plot_on_map()
+            f.savefig(files.dn_fig + "tidegauge_map.png")
```
Comment (jpolton) [inline unit_testing/test_tidegauge_methods.py:45]: More grid weirdness
tidegauge_map

File: unit_testing/test_isobath_contour_methods.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        check1 = np.allclose(
+            (
+                cont_f.data_cross_flow.normal_velocity_hpg
+                + cont_f.data_cross_flow.normal_velocity_spg
+                + cont_f.data_cross_flow.transport_across_AB_hpg
+                + cont_f.data_cross_flow.transport_across_AB_spg
+            ).sum(),
+            74.65002414,
+        )
+        self.assertTrue(check1, "check1")
```
Comment (jpolton) [inline unit_testing/test_isobath_contour_methods.py:86]: This fails:
```
======================================================================
ERROR: test_calculate_pressure_gradient_griven_flow (test_isobath_contour_methods.test_isobath_contour_methods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jeff/GitHub/COAsT/unit_testing/test_isobath_contour_methods.py", line 74, in test_calculate_pressure_gradient_griven_flow
    cont_f.calc_geostrophic_flow(
  File "/Users/jeff/GitHub/COAsT/coast/contour.py", line 726, in calc_geostrophic_flow
    self._update_cross_flow_vars(
  File "/Users/jeff/GitHub/COAsT/coast/contour.py", line 442, in _update_cross_flow_vars
    self.data_cross_flow[var] = tmp_var[:-1]
  File "/Users/jeff/opt/anaconda3/envs/coast_dev_38/lib/python3.8/site-packages/xarray/core/dataset.py", line 1564, in __setitem__
    self.update({key: value})
  File "/Users/jeff/opt/anaconda3/envs/coast_dev_38/lib/python3.8/site-packages/xarray/core/dataset.py", line 4253, in update
    merge_result = dataset_update_method(self, other)
  File "/Users/jeff/opt/anaconda3/envs/coast_dev_38/lib/python3.8/site-packages/xarray/core/merge.py", line 988, in dataset_update_method
    return merge_core(
  File "/Users/jeff/opt/anaconda3/envs/coast_dev_38/lib/python3.8/site-packages/xarray/core/merge.py", line 651, in merge_core
    raise MergeError(
xarray.core.merge.MergeError: unable to determine if these variables should be coordinates or not in the merged result: {'longitude'}

----------------------------------------------------------------------
Ran 24 tests in 176.747s

FAILED (errors=1)
```

**Comment by jpolton** [comment]
* Tidegauge also works for me.
* Profiles works (though there are lots of warnings):
```
test_compare_processed_profile_with_model (test_profile_methods.test_profile_methods) ... /Users/jeff/GitHub/COAsT/coast/logging_util.py:78: UserWarning: /Users/jeff/GitHub/COAsT/coast/profile.py.__init__.41: Object created but no file or directory specified: 
<coast.profile.Profile object at 0x7fd23acbb040>
  return warnings.warn(add_info(msg), *args, **kwargs)
/Users/jeff/GitHub/COAsT/coast/logging_util.py:78: UserWarning: /Users/jeff/GitHub/COAsT/coast/profile.py.__init__.41: Object created but no file or directory specified: 
<coast.profile.Profile object at 0x7fd23937a550>
  return warnings.warn(add_info(msg), *args, **kwargs)
/Users/jeff/GitHub/COAsT/coast/logging_util.py:78: UserWarning: /Users/jeff/GitHub/COAsT/coast/profile.py.__init__.41: Object created but no file or directory specified: 
<coast.profile.Profile object at 0x7fd23b66fb80>
  return warnings.warn(add_info(msg), *args, **kwargs)
/Users/jeff/GitHub/COAsT/coast/logging_util.py:78: UserWarning: /Users/jeff/GitHub/COAsT/coast/profile.py.__init__.41: Object created but no file or directory specified: 
<coast.profile.Profile object at 0x7fd23b6847c0>
  return warnings.warn(add_info(msg), *args, **kwargs)
/Users/jeff/GitHub/COAsT/coast/logging_util.py:78: UserWarning: /Users/jeff/GitHub/COAsT/coast/profile.py.__init__.41: Object created but no file or directory specified: 
<coast.profile.Profile object at 0x7fd239783c10>
  return warnings.warn(add_info(msg), *args, **kwargs)
/Users/jeff/GitHub/COAsT/coast/logging_util.py:78: UserWarning: /Users/jeff/GitHub/COAsT/coast/profile.py.__init__.41: Object created but no file or directory specified: 
<coast.profile.Profile object at 0x7fd22a2cca60>
  return warnings.warn(add_info(msg), *args, **kwargs)
ok
```

**Review by jpolton — APPROVED** [review]

File: unit_testing/test_tidegauge_methods.py (hunk 1 above)
Comment (davbyr) [reply unit_testing/test_tidegauge_methods.py:129]: I don't get this effect actually:
Screenshot 2022-03-25 at 11 38 15

**Pull request merged**